src/preprocess/query_generator.py

```diff
@@ -151,81 +174,420 @@ def generate_delete_datasets_rules(datasets, data_elements, org_units, org_unit_
+              " FROM program p INNER JOIN programstage ps ON p.programid = ps.programid INNER JOIN "
+              "programstagedataelement psde ON psde.programstageid = ps.programstageid WHERE  p.uid IN {})); \n"
+              .format(sql_event_program, sql_event_program))
+    if anonimize_coordinate:
+        write(f, "update programinstance as rand set geometry=null where programinstanceid "
+                 "in ( select psi.programinstanceid  from programstageinstance psi "
+                 "inner join programstage ps on ps.programstageid=psi.programstageid "
+                 "inner join program p on p.programid=ps.programid where p.uid in {});\n".format(sql_event_program))
+        write(f, "update programstageinstance as rand set geometry=null where programstageinstanceid "
+                 "in ( select psi.programstageinstanceid  from programstageinstance psi "
+                 "inner join programstage ps on ps.programstageid=psi.programstageid "
+                 "inner join program p on p.programid=ps.programid "
+                 "where p.uid in {});\n".format(sql_event_program))
+    if anonimize_org_units:
+        write(f, "update programinstance as rand set organisationunitid=(select organisationunitid "
+                 "from organisationunit where length(path)=(select length(path) from organisationunit where rand.organisationunitid= organisationunitid) ORDER BY RANDOM() limit 1) where programinstanceid "
+                 "in ( select psi.programinstanceid  from programstageinstance psi "
+                 "inner join programstage ps on ps.programstageid=psi.programstageid "
+                 "inner join program p on p.programid=ps.programid where p.uid in {});\n".format(sql_event_program))
+        write(f, "update programstageinstance as rand set organisationunitid="
+                 "(select organisationunitid from organisationunit where length(path)=(select length(path) from organisationunit where rand.organisationunitid= organisationunitid) ORDER BY RANDOM()"
+                 "limit 1) where programstageinstanceid "
+                 "in ( select psi.programstageinstanceid  from programstageinstance psi "
+                 "inner join programstage ps on ps.programstageid=psi.programstageid "
+                 "inner join program p on p.programid=ps.programid "
+                 "where p.uid in {});\n".format(sql_event_program))
+
+    if sql_data_elements != "":
+        if sql_data_elements != "":
+            for uid in data_elements:
+                write(f, "UPDATE programstageinstance SET    eventdatavalues = jsonb_set(eventdatavalues, "
+                         "'{" + uid + ",value}', concat(\'\',concat(concat(\'" + "\"" + "\',(select concat(\'Redacted:\',"
+                         "round(random()*dataelementid+programstageinstanceid))"
+                         ' FROM   dataelement WHERE  uid = \'sPadHOO4SQY\'),\'' + '\"' + '\'),\'\'))::jsonb)'
```

Comment:
This escaping can be hard to follow.`""" multi-line here, with free ' " quotes """`  should also work